offixsdk/src/main/java/org/aerogear/offixsdk/worker/OfflineMutationsWorker.kt

```diff
@@ -0,0 +1,115 @@
+ * Post firing the mutation, it is deleted from the db
+ */
+
+class OfflineMutationsWorker(context: Context, workerParameters: WorkerParameters) : Worker(context, workerParameters) {
+
+
+    private val getMutationDao = Application.database.mutationDao()
+    private val getUrlDao = Application.database.urlDao()
+
+    val TAG = javaClass.simpleName
+
+    override fun doWork(): Result {
+
+        Log.e(TAG, "In Worker")
+
+        //get the list of mutations
+
+        val listOfMutations = getMutationDao.getAllMutations()
+        Log.e(TAG, "Size of list : ${listOfMutations.size}")
+
+        //get the mutation one by one from database
+
+        listOfMutations.forEach { storedmutation ->
+
+            //Get the class name of the mutation to which it has to be mapped
+            val responseClassName = storedmutation.responseClassName
+
+            val classReflected: Class<*> = Class.forName(responseClassName)
+
+            //Get the constructor of the class, and as apollo generated classes have only one constructor, so take the first one.
+            val constructor = classReflected.constructors.first()
+
+            //Get the parameterTypes
+            val parameters = constructor.parameterTypes
+
+            val jsonValues = arrayListOf<Any>()
+
+            //Get the json object i.e the variables map given as input by the user
+            val jsonObj = storedmutation.valuemap
+
+            //Put all the json values into a list
+            val iter = jsonObj.keys()
+            iter.forEach { key ->
+                jsonValues.add(jsonObj.get(key))
+            }
+
+            Log.e("jsonValuesList ", " ${jsonValues.size}")
+
+            jsonValues.forEach {
+                Log.e("jsonValuesList : ", " $it")
+            }
+
+            //Check if the input parameter is of type Input<*>, if yes typecast it to be of the type Input<*>
+            parameters.forEachIndexed { index, clazz ->
+
+                Log.e("parameters : ", " ${clazz.name}")
+                if (clazz.name.equals("com.apollographql.apollo.api.Input")) {
+                    jsonValues[index] = Input.optional(jsonValues[index])
+                    Log.e("parameters **: ", " ${jsonValues[index].javaClass.name}")
+
+                }
+            }
+
+            //Make an object of mutation (done by reflection)
+            val obj = constructor.newInstance(*jsonValues.toArray())
+
+            //get the server url related to this mutation.
+            val currentUrl = getUrlDao.getAMutationUrl(storedmutation.SNo)
+
+            Log.e(TAG, "Current mutation serial number is: ${storedmutation.SNo}")
+            Log.e(TAG, "String url serial number is: ${currentUrl.SNo}")
+
+            // Make an apollo client which takes in mutation object and makes a call to server.
+            val customClient = ClientGenerator(currentUrl.urlString).getApolloClient()?.mutate(
```

Comment:
WOW. this is going to end really really bad! Having more than one client in the application is going to mess the cache. Apollo cache is not designed to work with more than one client. We can have data lost or overwritten or some operations not succeeding because of that. Apart from this creating client is very very heavy and consumes a lots of memory as cache is also kept in memory first and persisted periodically. 

Cache can easily grow towards 10-20mb of data for regular apps so we might have also performance problem related to that. 


Me thinking: Why we need new client every time. Can we just use existing one?

Reply:
@laurafitzgerald This is the most important comment and I think that this is very easy to fix.

Reply:
Sorry I meant @Lavanyagaur22

Reply:
@wtrocki did you ping me in error ere?

Reply:
Yep! Sorry

Reply:
getApolloClient() follows a singleton pattern of making an apollo client. Here we are just calling that and passing our custom mutation object. Aren't we?

Reply:
But can we use the actuall apollo client from app? I mean a single instance of apollo client per entire App that uses the library. We can do this many ways. 
What approach you think will be best here?

Reply:
Actually, we need apollo client in the library to make a server call in case the client is offline. So we will need to make a client in our library but we can create it in such a way that's a singleton. 
I am actually thinking of a way so that we can create an extension function on ApolloClient.Builder(). In this way, the user can modify the client accordingly (add things according to his/her needs) and we can also use the client when we make a server call after coming online from offline.

Reply:
yes.. but do we want to sacrifice users ability to create client as they wish. Client will probably require tons of stuff that will be specific to their domain. Things like authentication, logging and many requirements we cannot really predict may not be possible if we just create clinent on our own. 

Trick is to give people the freedom to use Apollo as they were using it and then just document how to add our stuff. 

For example, we can have separate classes like OfflineClient that will hold a reference to ApolloClient, have room initialized and many other helpers that will be required in the future.